community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/commands/expressions/Distance.scala

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v2_3.commands.expressions
+
+import org.neo4j.cypher.internal.compiler.v2_3.ExecutionContext
+import org.neo4j.cypher.internal.compiler.v2_3.pipes.QueryState
+import org.neo4j.cypher.internal.compiler.v2_3.symbols.SymbolTable
+import org.neo4j.cypher.internal.frontend.v2_3.symbols._
+
+
+case class Distance(p1: Expression, p2: Expression) extends Expression {
+
+  type POINT = Map[String, Any]
+
+  def apply(ctx: ExecutionContext)(implicit state: QueryState): Any = {
+    val points: Seq[POINT] = Seq(p1, p2).map {
+      case Identifier(name) => ctx(name) match {
+        case p: Point => p(ctx).asInstanceOf[POINT]
+        case m: POINT => m
+      }
+      case p: Point => p(ctx).asInstanceOf[POINT]
+    }
+    // TODO: Support better calculations, like https://en.wikipedia.org/wiki/Vincenty%27s_formulae
+    // TODO: Support more coordinate systems
+    if (HaversinCalculator.appliesTo(points(0), points(1)))
+      HaversinCalculator.distance(points(0), points(1))
```

Comment:
Perhaps these could be implemented with a function returning `Option[Double]`, called with `getOrElse` and in the else case throw the exception?